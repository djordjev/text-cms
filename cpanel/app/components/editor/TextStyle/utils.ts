import { BaseEditor, Editor, Element } from 'slate';
import { ReactEditor } from 'slate-react';

import { CustomElement } from '~/components/editor/VariationEditor/types';

const getSelectedStyle = (editor: BaseEditor & ReactEditor) => {
  const { selection } = editor;

  if (!selection) return;

  const [match] = Array.from(
    Editor.nodes<CustomElement>(editor, {
      at: selection,
      match: (n) =>
        !Editor.isEditor(n) &&
        Element.isElement(n) &&
        (n.type === 'paragraph' || n.type === 'heading')
    })
  );

  const [nodeType] = match;

  return nodeType;
};

export { getSelectedStyle };
