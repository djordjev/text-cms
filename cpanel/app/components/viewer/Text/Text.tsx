import classnames from 'classnames';
import { createElement, CSSProperties, FC, useMemo } from 'react';
import { Element } from 'slate';

import { CustomElement, CustomText } from '~/types/editor';

export interface TextProps {
  className?: string;
  text: string;
}

const Text: FC<TextProps> = (props) => {
  const { className, text } = props;

  // Setup
  const parsed = useMemo(() => JSON.parse(text), [text]);
  console.log(parsed);

  // Markup
  const renderLeaf = (child: CustomText, index: number) => {
    const { click, text } = child;

    const classes = classnames({
      'u-font-bold': child.bold,
      'u-italic': child.italic,
      'u-line-through': child.strikethrough,
      'u-underline': child.underline
    });

    const style: CSSProperties = {};
    if (child.color) {
      style.color = child.color;
    }

    if (click) {
      const { type } = click;

      const actionClasses = classnames(classes, {
        'u-btn u-btn-primary': type === 'primary',
        'u-btn u-btn-secondary': type === 'secondary',
        'u-link u-link-primary': type === 'link'
      });

      const onClick = () =>
        alert(`Clicked action element with action: ${click.action}`);

      const commonProps = {
        className: actionClasses,
        onClick,
        key: index,
        style
      };

      switch (type) {
        case 'link':
          return (
            <a
              {...commonProps}
              href={click.href}
              rel="noreferrer"
              target="_blank"
            >
              {text}
            </a>
          );
        case 'primary':
        case 'secondary':
          return <button {...commonProps}>{text}</button>;
        default:
          throw new Error('unknown type');
      }
    }

    return (
      <span className={classes} key={index} style={style}>
        {text}
      </span>
    );
  };

  const renderElement = (cnt: CustomElement | CustomText, index: number) => {
    if (!Element.isElement(cnt)) {
      return renderLeaf(cnt, index);
    }

    const { align, children, type } = cnt;

    let elem: string;

    switch (type) {
      case 'heading':
        elem = `h${cnt.level}`;
        break;
      case 'paragraph':
        elem = 'p';
        break;
      case 'template':
        elem = 'span';
        break;
      default:
        return null;
    }

    const blockClasses = classnames({
      'u-text-left': align === 'left',
      'u-text-center': align === 'center',
      'u-text-right': align === 'right',
      'u-text-4xl': elem === 'h1',
      'u-text-3xl': elem === 'h2',
      'u-text-lg': elem === 'h3',
      'u-text-md': elem === 'h4',
      'u-text-sm': elem === 'h5',
      'u-text-xs': elem === 'h6'
    });

    return createElement(
      elem,
      { className: blockClasses, key: index },
      ...children.map(renderElement)
    );
  };

  // Short-circuit
  if (!text || !Array.isArray(parsed)) return null;

  return <div className={className}>{parsed.map(renderElement)}</div>;
};

export { Text };
