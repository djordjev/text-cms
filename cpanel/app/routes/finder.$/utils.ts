import { CreateParam } from '~/types';

const getCreateOpenModal = (params: URLSearchParams): CreateParam | null => {
  const param = params.get('action');

  if (param === 'new-file' || param === 'new-folder') return param;

  return null;
};

export { getCreateOpenModal };
