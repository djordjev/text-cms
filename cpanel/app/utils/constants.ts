export const COLORS = {
  BLACK: '#000',
  PRIMARY: '#003EB3'
};

export const BUTTON_ACTION = '_action';
