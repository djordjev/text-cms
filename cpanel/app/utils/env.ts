const readEnvVariable = <T = string>(name: string) => {
  const vars = process.env;

  const val = vars[name];
  if (!val) throw new Error(`Variable ${name} not supplied`);

  return val as T;
};

export { readEnvVariable };
