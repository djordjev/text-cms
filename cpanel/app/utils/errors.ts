const buildErrorResponse = (status = 404, message = 'Not found') => {
  return new Response(message, { status, statusText: message });
};

export { buildErrorResponse };
