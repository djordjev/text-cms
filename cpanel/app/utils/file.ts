import { FileVariation } from '~/types';
import {
  ConditionAndChain,
  ConditionDescriptor,
  ConditionGroup,
  ConditionOperator
} from '~/types/condition';
import { CustomElement } from '~/types/editor';

const isHome = (name: string) => {
  return name.toLowerCase() === 'finder';
};

const isFile = (name: string) => {
  return name.endsWith('.txt');
};

const isFolder = (name: string) => {
  return !isFile(name);
};

const getPathFromSegments = (segments: string[], segment: string) => {
  const index = segments.indexOf(segment);

  if (index < 1) return '/finder';

  const mapped = segments.slice(0, index + 1).map((segment) => {
    if (isHome(segment)) return 'finder';

    return segment;
  });

  const joined = mapped.join('/');

  return `/${joined}`;
};

const parseDescriptor = (descriptor: unknown): ConditionDescriptor => {
  const msg = 'incorrect descriptor format';

  if (!Array.isArray(descriptor)) throw new Error(msg);
  if (descriptor.length !== 3) throw new Error(msg);

  const [name, operator, value] = descriptor;

  if (typeof name !== 'string') throw new Error(msg);
  if (typeof value !== 'string') throw new Error(msg);
  if (typeof operator !== 'string') throw new Error(msg);

  const operators = ['=', '>', '>=', '<', '<=', 'in'];
  const isValid = operators.includes(operator);

  if (!isValid) throw new Error(msg);

  return [name, operator as ConditionOperator, value];
};

const parseAndChain = (andChain: unknown): ConditionAndChain => {
  if (!Array.isArray(andChain)) throw new Error('incorrect and chain format');

  return andChain.map(parseDescriptor);
};

const parseConditions = (str?: string): ConditionGroup | null => {
  if (!str) return null;

  const parsed = JSON.parse(str);

  if (!Array.isArray(parsed)) return null;

  try {
    return parsed.map(parseAndChain);
  } catch {
    return null;
  }
};

const parseText = (text: string): CustomElement[] => {
  if (!text) throw new Error('invalid text format');

  const parsed = JSON.parse(text);

  if (!Array.isArray(parsed)) throw new Error('invalid text format');

  return parsed.map((v) => {
    const descendant: CustomElement = { ...v };

    return descendant;
  });
};

const parseFileVariations = (str: string): FileVariation[] | null => {
  if (!str) return null;

  const parsed = JSON.parse(str);

  if (!Array.isArray(parsed)) return null;

  const hasFields = parsed.every((e) => e.name && e.id && e.text);

  if (!hasFields) return null;

  return parsed.map((i) => {
    const variation: FileVariation = {
      condition: i.condition,
      id: i.id,
      name: i.name,
      text: i.text
    };

    return variation;
  });
};

export {
  getPathFromSegments,
  isFile,
  isFolder,
  isHome,
  parseConditions,
  parseFileVariations,
  parseText
};
