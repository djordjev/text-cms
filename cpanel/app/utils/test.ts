import { Descendant } from 'slate';

export const TEST_EDITOR: Descendant[] = [
  {
    type: 'heading',
    children: [
      {
        text: 'HEADING'
      }
    ],
    level: 3,
    align: 'center'
  },
  {
    type: 'paragraph',
    children: [
      {
        text: 'A '
      },
      {
        text: 'line',
        italic: true
      },
      {
        text: ' of '
      },
      {
        text: 'text',
        color: '#e41111',
        bold: true
      },
      {
        text: ' in a '
      },
      {
        text: 'paragraph',
        bold: true
      },
      {
        text: '.'
      }
    ]
  }
];

export const buildRequest = (data: Record<string, string>, method = 'POST') => {
  const form = new URLSearchParams(data);

  return new Request('https://www.test.local', { body: form, method });
};
