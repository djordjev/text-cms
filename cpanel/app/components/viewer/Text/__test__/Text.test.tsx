import { createRemixStub } from '@remix-run/testing';
import { render, screen } from '@testing-library/react';

import { TEST_EDITOR } from '~/utils/test';

import { Text, TextProps } from '../Text';

test('Text', () => {
  const props: TextProps = {
    text: JSON.stringify(TEST_EDITOR)
  };

  const createComponent = () => {
    const RemixStub = createRemixStub([
      {
        path: '/',
        Component: () => <Text {...props} />
      }
    ]);

    return <RemixStub />;
  };

  const { container } = render(createComponent());

  // Displays heading
  expect(screen.getByText('HEADING')).toBeInTheDocument();

  const matcher = (content: string, element: Element | null) => {
    return element?.textContent === 'A line of text in a paragraph.';
  };

  // Displays paragraph
  expect(screen.getByText(matcher)).toBeInTheDocument();

  // Sets up styling
  expect(screen.getByText(/line/)).toHaveClass('u-italic');
  expect(screen.getByText(/paragraph/)).toHaveClass('u-font-bold');
  expect(screen.getByText(/text/)).toHaveClass('u-font-bold');
  expect(screen.getByText(/text/)).toHaveAttribute('style');

  expect(container.childNodes).toMatchSnapshot();
});
