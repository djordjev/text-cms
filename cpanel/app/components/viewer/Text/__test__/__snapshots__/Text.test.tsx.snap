// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Text 1`] = `
NodeList [
  <div>
    <h3
      class="u-text-center u-text-lg"
    >
      <span
        class=""
      >
        HEADING
      </span>
    </h3>
    <p
      class=""
    >
      <span
        class=""
      >
        A 
      </span>
      <span
        class="u-italic"
      >
        line
      </span>
      <span
        class=""
      >
         of 
      </span>
      <span
        class="u-font-bold"
        style="color: rgb(228, 17, 17);"
      >
        text
      </span>
      <span
        class=""
      >
         in a 
      </span>
      <span
        class="u-font-bold"
      >
        paragraph
      </span>
      <span
        class=""
      >
        .
      </span>
    </p>
  </div>,
]
`;
